relays/ethereum/src/substrate_client.rs

```diff
@@ -128,91 +201,102 @@ pub async fn ethereum_header_known(
 			client.genesis_hash = Some(genesis_hash);
 			(client, genesis_hash)
 		}
 	};
-	let account_id = signer.public().as_array_ref().clone().into();
-	let (client, nonce) = next_account_index(client, account_id).await;
-	let nonce = match nonce {
-		Ok(nonce) => nonce,
-		Err(err) => return (client, Err(err)),
-	};
+	let account_id = params.signer.public().as_array_ref().clone().into();
+	let (client, nonce) = bail_on_error!(next_account_index(client, account_id).await);
 
-	let transaction = create_signed_submit_transaction(headers, &signer, nonce, genesis_hash);
-	let encoded_transaction = transaction.encode();
-	let (client, transaction_hash) = call_rpc(
+	let transaction = create_signed_submit_transaction(headers, &params.signer, nonce, genesis_hash);
+	let encoded_transaction = bail_on_arg_error!(
+		to_value(Bytes(transaction.encode())).map_err(|e| Error::RequestSerialization(e)),
+		client
+	);
+	let (client, _) = call_rpc(
 		client,
 		"author_submitExtrinsic",
-		Params::Array(vec![to_value(Bytes(encoded_transaction)).unwrap()]),
+		Params::Array(vec![encoded_transaction]),
+		|_| Ok(()),
 	)
 	.await;
 
-	(
-		client,
-		transaction_hash.map(|transaction_hash| (vec![transaction_hash], ids)),
-	)
+	(client, Ok(ids))
 }
 
 /// Submits unsigned Ethereum header to Substrate runtime.
 pub async fn submit_unsigned_ethereum_headers(
 	mut client: Client,
 	headers: Vec<QueuedEthereumHeader>,
-) -> (Client, Result<(Vec<TransactionHash>, Vec<EthereumHeaderId>), Error>) {
+) -> (Client, Result<Vec<EthereumHeaderId>, Error>) {
 	let ids = headers.iter().map(|header| header.id()).collect();
-	let mut transactions_hashes = Vec::new();
 	for header in headers {
 		let transaction = create_unsigned_submit_transaction(header);
 
-		let encoded_transaction = transaction.encode();
-		let (used_client, transaction_hash) = call_rpc(
-			client,
-			"author_submitExtrinsic",
-			Params::Array(vec![to_value(Bytes(encoded_transaction)).unwrap()]),
-		)
-		.await;
+		let encoded_transaction = bail_on_arg_error!(
+			to_value(Bytes(transaction.encode())).map_err(|e| Error::RequestSerialization(e)),
+			client
+		);
+		let (used_client, _) = bail_on_error!(
+			call_rpc(
+				client,
+				"author_submitExtrinsic",
+				Params::Array(vec![encoded_transaction]),
+				|_| Ok(()),
+			)
+			.await
+		);
 
 		client = used_client;
-		transactions_hashes.push(match transaction_hash {
-			Ok(transaction_hash) => transaction_hash,
-			Err(error) => return (client, Err(error)),
-		});
 	}
 
-	(client, Ok((transactions_hashes, ids)))
+	(client, Ok(ids))
+}
+
+/// Get GRANDPA authorities set at given block.
+pub async fn grandpa_authorities_set(client: Client, block: Hash) -> (Client, Result<Vec<u8>, Error>) {
+	let block = bail_on_arg_error!(to_value(block).map_err(|e| Error::RequestSerialization(e)), client);
+	call_rpc(
+		client,
+		"state_call",
+		Params::Array(vec![
+			serde_json::Value::String("GrandpaApi_grandpa_authorities".into()),
```

Comment:
Let's also please move all strings to constants placed in the top part of the module.

Reply:
See: https://github.com/paritytech/parity-bridges-common/issues/75